cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStoreProvider.java

```diff
@@ -0,0 +1,543 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.security.securestore;
+
+import co.cask.cdap.api.security.securestore.SecureStoreMetadata;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Configuration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.internal.security.SecureStoreProvider;
+import com.google.inject.Singleton;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.Nullable;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * File based implementation of secure store. Uses Java JCEKS based keystore.
+ */
+@Singleton
+class FileSecureStoreProvider implements SecureStoreProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(FileSecureStoreProvider.class);
+
+  private static final String SCHEME_NAME = "jceks";
+  /*
+   This is used to create the key for storing metadata in the keystore. Changing this will make the
+   metadata for existing entries unreachable.
+   */
+  private static final String METADATA_SUFFIX = "_metadata";
+  private static final String SECURE_STORE_DEFAULT_FILE_PATH = "/tmp";
+  private static final String SECURE_STORE_DEFAULT_FILE_NAME = "securestore";
+  private static final char[] SECURE_STORE_DEFAULT_PASSWORD = "cdapsecret".toCharArray();
+  /*
+   Java Keystore needs an algorithm name to store a key, it is not used for any checks, only stored,
+   since we are not handling the encryption we don't care about this.
+  */
+  private static final String ALGORITHM_PROXY = "none";
+
+  /*
+   A cache for metadata.
+   The assumption is that the number of entries won't be large enough to worry about memory.
+  */
+  private final Map<String, SecureStoreMetadata> cache = new HashMap<>();
+  private final char[] password;
+  private final Path path;
+  private final Lock readLock;
+  private final Lock writeLock;
+  private final CConfiguration cConf;
+
+  private KeyStore keyStore;
```

Comment:
final. You can do `locateKeystore` from constructor and have the constructor `throws IOException`.